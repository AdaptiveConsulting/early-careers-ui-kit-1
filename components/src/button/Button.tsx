import React from "react"
import "./button.css"

interface ButtonProps {
  /**
   * on light or on dark background?
   */
  onLightBg?: boolean
  /**
   * Type of button?
   */
  buttonType?: string
  // /**
  //  * Is this the principal call to action on the page?
  //  */
  // primary?: boolean;
  /**
   * What background color to use
   */
  backgroundColor?: string
  // /**
  //  * How large should the button be?
  //  */
  // size?: 'small' | 'medium' | 'large';
  /**
   * Button contents
   */
  label: string
  /**
   * Optional click handler
   */
  onClick?: () => void
}

/**
 * Primary UI component for user interaction
 */
export const Button = ({
  buttonType,
  onLightBg,
  backgroundColor,
  label,
  ...props
}: ButtonProps) => {
  let base, hover, active, focus
  if (buttonType === "secondary" && onLightBg) {
    base =
      "font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-dark outline outline-1 outline-offset-[-1] outline-primary-dark "
    hover = "hover:bg-primary-dark hover:text-white hover:outline-none"
    active =
      "active:bg-transparent active:text-primary-dark active:outline active:outline-1 active:outline-offset-[-1] active:outline-primary-dark active:ring-0 "
    focus = "focus:ring-2 focus:ring-warning-main focus:ring-offset-2 "
  } else if (buttonType === "secondary" && !onLightBg) {
    base =
      "font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-light outline outline-1 outline-offset-[-1] outline-primary-light "
    hover = "hover:bg-primary-light hover:text-gray-800 "
    active =
      "active:bg-transparent active:text-primary-dark active:outline active:outline-1 active:outline-offset-[-1] active:outline-primary-main active:ring-0 "
    focus = "focus:ring-2 focus:ring-warning-light focus:ring-offset-2 "
  } else if (buttonType === "tertiary" && onLightBg) {
    base =
      "font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-dark outline-none outline-offset-0 no-underline "
    hover = "hover:underline "
    active =
      "active:bg-transparent active:text-primary-dark active:outline active:outline-1 active:outline-offset-[-1] active:outline-primary-dark active:ring-0 active:no-underline "
    focus =
      "focus:ring-2 focus:ring-warning-main focus:ring-offset-2 focus:bg-transparent focus:text-primary-dark focus:outline focus:outline-1 focus:outline-offset-[-1] focus:outline-primary-dark focus:no-underline"
  } else if (buttonType === "tertiary" && !onLightBg) {
    base =
      "font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-light outline-none outline-offset-0 no-underline "
    hover = "hover:underline "
    active =
      "active:bg-transparent active:text-primary-light active:outline active:outline-1 active:outline-offset-[-1] active:outline-primary-light active:ring-0 active:no-underline "
    focus =
      "focus:ring-2 focus:ring-warning-main focus:ring-offset-2 focus:bg-transparent focus:text-primary-light focus:outline focus:outline-1 focus:outline-offset-[-1] focus:outline-primary-light focus:no-underline"
  } else {
    //Primary
    base =
      "font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-primary-light text-black outline-none outline-offset-0 "
    hover = "hover:bg-primary-main "
    active =
      "active:outline-1 active:outline-offset-[-1] active:outline-primary-main active:bg-primary-light active:ring-0 "
    focus = "focus:ring-2 focus:ring-warning-main focus:ring-offset-2 "
  }

  return (
    <button
      type="button"
      className={base + hover + active + focus}
      style={{ backgroundColor }}
      {...props}
    >
      {label}
    </button>
  )
}
