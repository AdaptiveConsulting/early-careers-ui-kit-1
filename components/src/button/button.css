@tailwind base;
@tailwind components;
@tailwind utilities;

.hello {
    color: red;
    background-color: deeppink;
}