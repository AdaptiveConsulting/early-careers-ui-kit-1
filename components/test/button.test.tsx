import { assert, expect, test } from "vitest"
import renderer from "react-test-renderer"
import { Button } from "../src/button/Button"
function toJson(component: renderer.ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
}

test("Primary Button", () => {
  const component = renderer.create(<Button label="BUTTON TEXT" />)
  const tree = toJson(component)

  expect(tree.props.className).toBe(
    "btn-base bg-primary-light text-black outline-none outline-offset-0 hover:bg-primary-main active:outline-1 active:outline-offset-[-1] active:outline-primary-main active:bg-primary-light active:ring-0 active:ring-offset-0 focus:ring-2 focus:ring-warning-main focus:ring-offset-2 focus:ring-offset-parent",
  )
})

test("Secondary Button On Light Background", () => {
  const component = renderer.create(
    <Button label="BUTTON TEXT" variant="secondary" isDarkTheme={true} />,
  )
  const tree = toJson(component)

  expect(tree.props.className).toBe(
    "btn-base btn-secondary text-primary-light outline-primary-light hover:outline-none hover:bg-primary-light hover:text-gray-800 active:bg-transparent active:outline active:outline-1 active:outline-offset-[-1] active:ring-offset-0 active:ring-0 active:outline-primary-main active:text-primary-light focus:ring-2 focus:ring-offset-2 active:bg-transparent focus:ring-warning-light ",
  )
})

test("Secondary Button On Dark Background", () => {
  const component = renderer.create(
    <Button label="BUTTON TEXT" variant="secondary" isDarkTheme={false} />,
  )
  const tree = toJson(component)

  expect(tree.props.className).toBe(
    "btn-base btn-secondary text-primary-dark outline-primary-dark hover:outline-none hover:bg-primary-dark hover:text-common-white active:bg-transparent active:outline active:outline-1 active:outline-offset-[-1] active:ring-offset-0 active:ring-0 active:outline-primary-dark active:text-primary-dark focus:ring-2 focus:ring-offset-2 active:bg-transparent focus:ring-warning-main ",
  )
})

test("Tertiary Button On Light Background", () => {
  const component = renderer.create(
    <Button label="BUTTON TEXT" variant="tertiary" isDarkTheme={true} />,
  )
  const tree = toJson(component)
  expect(tree.props.className).toBe(
    "btn-base btn-tertiary text-primary-dark hover:underline active:text-primary-dark active:outline-primary-dark active:bg-transparent active:outline active:outline-1 active:outline-offset-[-1] active:ring-0 active:ring-offset-0 active:no-underline focus:ring-warning-light focus:outline-primary-dark focus:text-primary-dark focus:ring-2 focus:ring-offset-2 focus:bg-transparent focus:outline focus:outline-1 focus:outline-offset-[-1] focus:no-underline",
  )
})

test("Tertiary Button On Dark Background", () => {
  const component = renderer.create(
    <Button label="BUTTON TEXT" variant="tertiary" isDarkTheme={false} />,
  )

  const tree = toJson(component)
  expect(tree.props.className).toBe(
    "btn-base btn-tertiary text-primary-light hover:underline active:text-primary-light active:outline-primary-light active:bg-transparent active:outline active:outline-1 active:outline-offset-[-1] active:ring-0 active:ring-offset-0 active:no-underline focus:ring-warning-main focus:outline-primary-light focus:text-primary-light focus:ring-2 focus:ring-offset-2 focus:bg-transparent focus:outline focus:outline-1 focus:outline-offset-[-1] focus:no-underline",
  )
})

test("Disabled", () => {
  const component = renderer.create(<Button label="BUTTON TEXT" disabled />)

  const tree = toJson(component)

  expect(tree.props.className).toBe("btn-base bg-grey-200 text-grey-500")
})
