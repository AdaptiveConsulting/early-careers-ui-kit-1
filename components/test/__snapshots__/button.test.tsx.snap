// Vitest Snapshot v1

exports[`Primary Button 1`] = `
<button
  className="font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-primary-light text-black outline-none outline-offset-0 hover:bg-primary-main active:outline-1 active:outline-offset--1 active:outline-primary-main active:bg-primary-light active:ring-0 active:ring-offset-0 focus:ring-2 focus:ring-warning-main focus:ring-offset-2 focus:ring-offset-parent"
  style={
    {
      "backgroundColor": undefined,
    }
  }
  type="button"
>
  BUTTON TEXT
</button>
`;

exports[`Secondary Button On Dark Background 1`] = `
<button
  className="font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-light outline outline-1 outline-offset--1 outline-primary-light hover:bg-primary-light hover:text-gray-800 hover:outline-none active:bg-transparent active:outline active:outline-1 active:-outline-offset-1 active:ring-offset-0 active:ring-0 active:outline-primary-main active:text-primary-light focus:ring-warning-light focus:ring-2 focus:ring-offset-2 active:bg-transparent "
  style={
    {
      "backgroundColor": undefined,
    }
  }
  type="button"
>
  BUTTON TEXT
</button>
`;

exports[`Secondary Button On Light Background 1`] = `
<button
  className="font-roboto text-[11px] rounded-full w-auto h-21px px-4 bg-transparent text-primary-dark outline outline-1 outline-offset--1 outline-primary-dark hover:bg-primary-dark hover:text-white hover:outline-none active:bg-transparent active:outline active:outline-1 active:-outline-offset-1 active:ring-offset-0 active:ring-0 active:outline-primary-dark active:text-primary-dark focus:ring-warning-main focus:ring-2 focus:ring-offset-2 active:bg-transparent "
  style={
    {
      "backgroundColor": undefined,
    }
  }
  type="button"
>
  BUTTON TEXT
</button>
`;

exports[`Tertiary Button On Dark Background 1`] = `
<button
  className="font-roboto text-[11px] rounded-full w-auto h-21px px-4 text-primary-light bg-transparent outline-none outline-offset-0 no-underline hover:underline active:text-primary-light active:outline-primary-light active:bg-transparent active:outline active:outline-1 active:outline-offset--1 active:ring-0 active:ring-offset-0 active:no-underline focus:ring-warning-light focus:outline-primary-light focus:text-primary-light focus:ring-2 focus:ring-offset-2 focus:bg-transparent focus:outline focus:outline-1 focus:outline-offset--1 focus:no-underline"
  style={
    {
      "backgroundColor": undefined,
    }
  }
  type="button"
>
  BUTTON TEXT
</button>
`;

exports[`Tertiary Button On Light Background 1`] = `
<button
  className="font-roboto text-[11px] rounded-full w-auto h-21px px-4 text-primary-dark bg-transparent outline-none outline-offset-0 no-underline hover:underline active:text-primary-dark active:outline-primary-dark active:bg-transparent active:outline active:outline-1 active:outline-offset--1 active:ring-0 active:ring-offset-0 active:no-underline focus:ring-warning-main focus:outline-primary-dark focus:text-primary-dark focus:ring-2 focus:ring-offset-2 focus:bg-transparent focus:outline focus:outline-1 focus:outline-offset--1 focus:no-underline"
  style={
    {
      "backgroundColor": undefined,
    }
  }
  type="button"
>
  BUTTON TEXT
</button>
`;
